core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/DefaultCompressionAlgorithm.java

```diff
@@ -0,0 +1,355 @@
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.Maps;
+
+/**
+ * Compression algorithms. There is a static initializer, below the values defined in the
+ * enumeration, that calls the initializer of all defined codecs within
+ * {@link DefaultCompressionAlgorithm}. This promotes a model of the following call graph of
+ * initialization by the static initializer, followed by calls to {@link #getCodec()},
+ * {@link #createCompressionStream(OutputStream, Compressor, int)}, and
+ * {@link #createDecompressionStream(InputStream, Decompressor, int)}. In some cases, the
+ * compression and decompression call methods will include a different buffer size for the stream.
+ * Note that if the compressed buffer size requested in these calls is zero, we will not set the
+ * buffer size for that algorithm. Instead, we will use the default within the codec.
+ * <p>
+ * The buffer size is configured in the Codec by way of a Hadoop {@link Configuration} reference.
+ * One approach may be to use the same Configuration object, but when calls are made to
+ * {@code createCompressionStream} and {@code createDecompressionStream} with non default buffer
+ * sizes, the configuration object must be changed. In this case, concurrent calls to
+ * {@code createCompressionStream} and {@code createDecompressionStream} would mutate the
+ * configuration object beneath each other, requiring synchronization to avoid undesirable activity
+ * via co-modification. To avoid synchronization entirely, we will create Codecs with their own
+ * Configuration object and cache them for re-use. A default codec will be statically created, as
+ * mentioned above to ensure we always have a codec available at loader initialization.
+ * <p>
+ * There is a Guava cache defined within Algorithm that allows us to cache Codecs for re-use. Since
+ * they will have their own configuration object and thus do not need to be mutable, there is no
+ * concern for using them concurrently; however, the Guava cache exists to ensure a maximal size of
+ * the cache and efficient and concurrent read/write access to the cache itself.
+ * <p>
+ * To provide Algorithm specific details and to describe what is in code:
+ * <p>
+ * LZO will always have the default LZO codec because the buffer size is never overridden within it.
+ * <p>
+ * LZ4 will always have the default LZ4 codec because the buffer size is never overridden within it.
+ * <p>
+ * GZ will use the default GZ codec for the compression stream, but can potentially use a different
+ * codec instance for the decompression stream if the requested buffer size does not match the
+ * default GZ buffer size of 32k.
+ * <p>
+ * Snappy will use the default Snappy codec with the default buffer size of 64k for the compression
+ * stream, but will use a cached codec if the buffer size differs from the default.
+ */
+public class DefaultCompressionAlgorithm extends Configured {
```

Comment:
I think the Javadoc needs to be updated. I am still confused by this class. It doesn't really make sense to call it a "Default" class since it really isn't an implementation of an interface or abstract class. It extends the Hadoop type, which I am not sure if need or not.  Then it just sets up all the different types. Maybe this should be called `CompressionHelper`? It might make sense to split the Hadoop configuration type from the compression classes so you would have a `CompressionHelper` type, which sets up all the compression stuff and then the hadoop configured type to hold what is pulled from the hadoop configuration. The enum was just called `Algorithm` before so maybe `CompressionAlgorithm` makes more sense.

Sorry, I know your changes are just modifications to what was already there but it was confusing before so I am just trying to reduce the confusion if we can.

Reply:
I talked with @dlmarion and he helped clear up some of the confusion. The code was confusing before so I think we can clean up the internal stuff as a follow on.